ddtrace/contrib/grpc/client_interceptor.py

```diff
@@ -1,61 +1,82 @@
+import collections
 import grpc
 
-from ddtrace import Pin
-from .propagation import inject_span
+from ddtrace import config
+from ...propagation.http import HTTPPropagator
 from ...constants import ANALYTICS_SAMPLE_RATE_KEY
-from ...settings import config
 
+# DEV: Follows Python interceptors RFC laid out in
+# https://github.com/grpc/proposal/blob/master/L13-python-interceptors.md
 
-class GrpcClientInterceptor(
-        grpc.UnaryUnaryClientInterceptor, grpc.UnaryStreamClientInterceptor,
-        grpc.StreamUnaryClientInterceptor, grpc.StreamStreamClientInterceptor):
-    """Intercept calls on a channel. It creates span as well as doing the propagation
-    Derived from https://github.com/grpc/grpc/blob/d0cb61eada9d270b9043ec866b55c88617d362be/examples/python/interceptors/headers/generic_client_interceptor.py#L19
-    """  # noqa
-
-    def __init__(self, host, port):
-        self._pin = Pin.get_from(grpc)
-        self._host = host
-        self._port = port
-
-    def _start_span(self, method):
-        span = self._pin.tracer.trace('grpc.client', span_type='grpc', service=self._pin.service, resource=method)
-        span.set_tag('grpc.host', self._host)
-        if (self._port is not None):
-            span.set_tag('grpc.port', self._port)
-        if self._pin.tags:
-            span.set_tags(self._pin.tags)
-        # set analytics sample rate
-        span.set_tag(
-            ANALYTICS_SAMPLE_RATE_KEY,
-            config.grpc.get_analytics_sample_rate()
-        )
-        return span
 
-    def intercept_unary_unary(self, continuation, client_call_details, request):
-        return self.intercept_unary_stream(continuation, client_call_details, request)
+def create_client_interceptor(pin):
+    def interceptor_function(continuation, client_call_details,
+                             request_or_iterator):
+        if not pin.enabled:
+            response = continuation(client_call_details, request_or_iterator)
+            return response
+
+        with pin.tracer.trace(
+                '{}.client'.format(pin.app),
+                span_type=config.grpc.span_type,
+                service=pin.service,
+                resource=client_call_details.method
+        ) as span:
+            if pin.tags:
+                span.set_tags(pin.tags)
+
+            span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.grpc.get_analytics_sample_rate())
+
+            headers = {}
+            if config.grpc.distributed_tracing_enabled:
+                propagator = HTTPPropagator()
+                propagator.inject(span.context, headers)
+
+            metadata = []
+            if client_call_details.metadata is not None:
+                metadata = list(client_call_details.metadata)
+            metadata.extend(headers.items())
+
+            client_call_details = _ClientCallDetails(
```

Comment:
Ah, we can't do `client_call_details.metadata = metadata` right? This is a `NamedTuple`, IIRC?

Reply:
Right, we defined it as `collections.namedtuple` with those attributes. The implementation here does not mutate the intercepted `client_call_details` and instead creates a new one to pass along.

Reply:
yeah, we can't anyways since it is a `namedtuple`... we _could_ if it wasn't `None`, since you can mutate the list in place, just can't do `client_call_details.metadata = `